platform-sdk/docs/proposals/queuestate-fcqueue-to-virtualmap/proposal.md

```diff
@@ -0,0 +1,181 @@
+# Design proposal: queue states - migrate from FCQueue to VirtualMap
+
+---
+
+## Summary
+
+This design proposal covers details how to re-implement Queue state APIs on top of `VirtualMap`
+rather than `FCQueue` as it is today. The main goal is to have applications like Hedera services
+unaffected, and have data migration fast enough to be done during software update cycle.
+
+## Purpose and context
+
+State APIs is an abstraction layer on top of the merkle state tree. This layer provides methods
+and classes to work with ledger data regardless of how it's stored in memory and on disk. There
+are 3 different state types supported:
+
+* Singleton states: store a single value
+* Queue states: store multiple values with queue-like APIs
+* Key/value states: store multiple values with map-like APIs
+
+For production use cases, queue states are implemented using a single `QueueNode` merkle node, which
+is an internal merkle node with two children: a label and a queue. The label is a `StringLeaf`, and
+the queue is an `FCQueue`.
+
+`FCQueue` is an in-memory implementation of a queue. It provides all APIs expected from a queue,
+like to add an element to the head of the queue or to poll an element at the tail of the queue. It
+also provides efficient implementation of hashCode(). This is critical, since queues may contain
+lots of elements, and queue hashes must be computed fast.
+
+There are several drawbacks in `FCQueue`, too:
+
+* All data is in memory
+* When a state snapshot is taken, all queue elements are serialized to disk one by one
+* During reconnects, the whole queue is serialized and sent to the learner
+
+Instead, all the queue APIs can be implemented on top of a `VirtualMap`. It's more memory friendly,
+optimizied for reconnects and snapshots.
+
+## Architecture
+
+### Diagrams
+
+|              Current              |                Proposal                |
+|-----------------------------------|----------------------------------------|
+| ![Current](architecture-0.54.svg) | ![Proposal](architecture-proposal.svg) |
+
+### Java classes
+
+No changes to `ReadableQueueState` and `WritableQueueState` interfaces.
+
+No changes to `ReadableQueueStateImpl` and `WritableQueueStateImpl` classes.
+
+`QueueNode`:
+
+* No longer marked as `@Deprecated`
```

Comment:
to clarify, `QueueNode` will be undeprecated but `FCQueue` will now be deprecated?

Reply:
+0.8

Reply:
Correct